### ISSUE TITLE
Repeating a cached query with a very large result fails

### ISSUE THREAD

**Issue by janiemi** [spraakbanken/korp-backend#7]
If you make a query with a very large result (at least with 280 million hits or more), repeating the query before the cache is cleared fails with an error. For example, performing the search
https://ws.spraakbanken.gu.se/ws/korp/v8/query?default_context=1%20sentence&corpus=FAMILJELIV-ALLMANNA-SAMHALLE&cqp=[]&default_within=sentence&cache=true&debug=true
twice (or more) in a row results in the following error on the second time:

```
{
"ERROR": {
"type": "CQPError",
"value": "ERROR: File length of subcorpus is <= 0",
"traceback": [
"Traceback (most recent call last):",
"  File \"/home/fkkorp/korp-backend/v8/korp.py\", line 223, in error_catcher",
"    g(*pargs, **kwargs)",
"  File \"/home/fkkorp/korp-backend/v8/korp.py\", line 213, in f",
"    for response in generator(args, *pargs, **kwargs):",
"  File \"/home/fkkorp/korp-backend/v8/korp.py\", line 617, in query",
"    start=hits[0], end=hits[1], **queryparams)",
"  File \"/home/fkkorp/korp-backend/v8/korp.py\", line 1198, in query_and_parse",
"    random_seed, no_results, expand_prequeries, free_search, use_cache)",
"  File \"/home/fkkorp/korp-backend/v8/korp.py\", line 1006, in query_corpus",
"    next(lines)",
"  File \"/home/fkkorp/korp-backend/v8/korp.py\", line 3218, in run_cqp",
"    raise CQPError(error)",
"korp.CQPError: ERROR: File length of subcorpus is <= 0"
]
},
"time": 0.6492030620574951
}
```

In the Korp frontend, this means that the first such query succeeds, but for example, changing KWIC page results in an error.

The error occurs when CQP tries to read the previously saved named query results (NQR) file. It seems that CQP can write an NQR file larger than 2 GiB but it cannot read such a file if its size is between 2 and 4 GiB. When CQP implicitly converts the file size to an `int` on a system with 32-bit `int`s, such a size is converted to a negative number, and CQP checks that the size of an NQR file is positive before reading it.

I reported this CWB bug or limitation via the CWB bug tracker: https://sourceforge.net/p/cwb/bugs/75/.

I don’t know if there is any way to avoid writing such a large NQR file, as you cannot make CQP commands conditional. Thus, as long as the limitation is present in CQP, the workaround I’d see possible in the Korp backend would be to check the size of the saved NQR file after executing `cqp` and to remove it if it is 2 GiB or larger. (The maximum size could also be configurable.) Unless you do that first, I’ll try to implement such a workaround at some point and make a pull request.

**Comment by janiemi** [comment]
The [CWB bug](https://sourceforge.net/p/cwb/bugs/75/) was fixed faster than I expected, in [revision r1708](https://sourceforge.net/p/cwb/code/1708/), so if you use that revision or later, you are not affected by this issue. However, I might still be inclined to add a configuration option to the Korp backend for immediately removing NQR files larger than a certain threshold, so I won’t close this ticket quite yet.

**Comment by MartinHammarstedt** [comment]
Sorry for not reacting to this issue sooner. Very convenient that they solved the issue so quickly!

I agree that such an option would be useful in Korp anyway, and it is now implemented in commit 8cebe4d. You can now use the `CACHE_MAX_QUERY_DATA` setting to set a max size for cached query results.
